Objects/dictobject.c

```diff
@@ -5423,23 +5423,23 @@ int
 _PyObject_StoreInstanceAttribute(PyObject *obj, PyDictValues *values,
                               PyObject *name, PyObject *value)
 {
-    assert(PyUnicode_CheckExact(name));
     PyDictKeysObject *keys = CACHED_KEYS(Py_TYPE(obj));
     assert(keys != NULL);
     assert(values != NULL);
     assert(Py_TYPE(obj)->tp_flags & Py_TPFLAGS_MANAGED_DICT);
-    Py_ssize_t ix = insert_into_dictkeys(keys, name);
+    Py_ssize_t ix = DKIX_EMPTY;
+    if (PyUnicode_CheckExact(name)) {
+        ix = insert_into_dictkeys(keys, name);
+    }
     if (ix == DKIX_EMPTY) {
-        if (value == NULL) {
-            PyErr_SetObject(PyExc_AttributeError, name);
-            return -1;
-        }
 #ifdef Py_STATS
-        if (shared_keys_usable_size(keys) == SHARED_KEYS_MAX_SIZE) {
-            OBJECT_STAT_INC(dict_materialized_too_big);
-        }
-        else {
-            OBJECT_STAT_INC(dict_materialized_new_key);
+        if (PyUnicode_CheckExact(name)) {
+            if (shared_keys_usable_size(keys) == SHARED_KEYS_MAX_SIZE) {
+                OBJECT_STAT_INC(dict_materialized_too_big);
+            }
+            else {
+                OBJECT_STAT_INC(dict_materialized_new_key);
+            }
```

Comment:
When name is not unicode, no stat is incremented?

Reply:
Added

Reply:
My thought was just incrementing dict_materialized_new_key.
Since str subclass is too rare, I am not sure it's worth enough to add new stat key.

But you are the top user of the stat. Please do what you think best.